test/Npgsql.Tests/Types/CompositeTests.cs

```diff
@@ -792,5 +792,77 @@ class CompositeWithoutSetter
         {
             public int Value { get; }
         }
+
+        [Test]
+        public void ThrowsOnReadWithoutPublicConstructor()
```

Comment:
[EF Core doesn't require constructors to be public](https://docs.microsoft.com/en-us/ef/core/modeling/constructors). I think that's a good idea: it allows users to prevent application code from using a constructor, but allows the data layer to materialize those objects.

I think we should do the same and allow non-public constructors - any reason not to?

Reply:
We use only public properties and fields, so only public constructors. Maybe there should be options to set up the handler, but this is another story.

Reply:
FYI the EF Core behavior is to not look at non-public properties and fields by default, put to look at non-public constructors. This makes sense to me as a default behavior. If a non-public constructor exists whose parameters match perfectly, should we really not use it?

Reply:
Let's start with publics and ask the community for feedback. We always are able to tune this in some patch release, maybe in the first. Otherwise, it would be a breaking change.

/cc @NinoFloris @austindrenski

Reply:
I'm OK with this.